crates/database/src/app_data.rs

```diff
@@ -0,0 +1,83 @@
+use {crate::AppId, sqlx::PgConnection};
+
+/// Tries to associate the contract app data with the full app data.
+///
+/// If this contract app data already existed then the existing full app data is
+/// returned.
+///
+/// If this contract did not already exist, it is inserted and the passed in
+/// full app data is returned.
+pub async fn insert(
+    ex: &mut PgConnection,
+    contract_app_data: &AppId,
+    full_app_data: &[u8],
+) -> Result<Vec<u8>, sqlx::Error> {
+    const QUERY: &str = r#"
+WITH insertion AS (
+    INSERT INTO app_data (contract_app_data, full_app_data)
+    VALUES ($1, $2)
+    -- returns null on conflict
+    ON CONFLICT DO NOTHING
+    -- returns inserted value without conflict
+    RETURNING full_app_data
+)
+SELECT COALESCE(
+    (SELECT full_app_data FROM insertion),
+    (SELECT full_app_data FROM app_data WHERE contract_app_data = $1)
+)
+;"#;
+    sqlx::query_scalar(QUERY)
+        .bind(contract_app_data)
+        .bind(full_app_data)
+        .fetch_one(ex)
+        .await
+}
+
+pub async fn fetch(
+    ex: &mut PgConnection,
+    contract_app_data: &AppId,
+) -> Result<Option<Vec<u8>>, sqlx::Error> {
+    const QUERY: &str = r#"
+SELECT full_app_data
+FROM app_data
+WHERE contract_app_data = $1
+;"#;
+    sqlx::query_scalar(QUERY)
+        .bind(contract_app_data)
+        .fetch_optional(ex)
+        .await
+}
+
+#[cfg(test)]
+mod tests {
+    use {super::*, crate::byte_array::ByteArray, sqlx::Connection};
+
+    #[tokio::test]
+    #[ignore]
+    async fn postgres_app_data() {
+        let mut db = PgConnection::connect("postgresql://").await.unwrap();
+        let mut tx = db.begin().await.unwrap();
+        crate::clear_DANGER_(&mut tx).await.unwrap();
+        tx.commit().await.unwrap();
```

Comment:
I'm a bit surprised to see the `commit` here. Not that is makes a difference it just looks like all the other DB tests never commit the tx.
Does **not** committing the tx maybe lead to a deadlock or some other problem?

Reply:
This test is special in a way I didn't document. Transactions fail when one of their query execution fails. In this test I'm testing a failing query and I keep using the database afterwards. If that happened in a transaction then the queries after the first failure would all fail because the transaction as a whole failed.